### ISSUE TITLE
excessive memory consumption while probing larger apps

### ISSUE THREAD

**Issue by stanfordcox** [dyninst/dyninst#756]
Using the systemtap dyninst backend, it  runs into trouble while probling a larger app with many large shared objects.

Probing an app like firefox or emacs that pulls in many large shared objects can frequently get a SIGTERM due to excessive process memory use while doing the initial bpatch.processAttach

To reproduce without use of stapdyn download:
 https://scox.fedorapeople.org/dyninst/mutator-pp.cpp
 % mutator-pp -dynamic -count-snippet -f main -p $(pgrep -n emacs)
 (s/emacs/firefox)
The program calls mallinfo.  Notice in particular uordblks/fordblks.  At begin and end of main I see:
Total allocated space (uordblks):             689744
Total free space (fordblks):                      170416
----------------------------------------------------------
Total allocated space (uordblks):      1603713184
Total free space (fordblks):                     2045792

So we have 1,603,023,440 bytes that are not freed.  Given a 4kb page that is 391363 pages; so a bunch of memory that stays allocated.

Expected would be for these to work:
 %  mutator-pp -dynamic -count-snippet -f main -p $(pgrep -n emacs)
 % mutator-pp -dynamic -count-snippet -f read_event_from_main_queue -p $(pgrep -n emacs)
 % mutator-pp -dynamic -count-snippet -f SplitAtCommas -p $(pgrep -n firefox)
-f FUNCTION-NAME will vary of course depending on version of emacs and firefox

This should be os independent and I can verify on rhel7 and rhel8 but I am currently testing this on:
kernel-5.6.13-300.fc32.x86_64
glibc-2.31-2.fc32.x86_64
gcc-c++-10.1.1-1.fc32.x86_64
boost-1.69.0-18.fc32.x86_64
tbb-2020.2-1.fc32.x86_64

It would probably be difficult to do the equivalent but gdb does piecemeal symbol analysis,  admittingly a much easier problem.  From gdb.info "the 'symbol-file' command does not normally
read the symbol table in full right away.  Instead, it scans the
symbol table quickly to find which source files and which symbols
are present.  The details are read later, one source file at a
time, as they are needed."

**Comment by mxz297** [comment]
Stan and I have been working the memory usage problem for several days.

I ran a heap memory profiler (https://github.com/KDE/heaptrack) with Stan's reproducer in a Fedora 31 docker image. The summary output are:

```
peak heap memory consumption: 14.50GB
peak RSS (including heaptrack overhead): 17.48GB
total memory leaked: 14.50GB
```

Basically, all memory is leaked. I am not totally surprised by this huge amount of leaks because Dyninst does not carefully free memory before exiting. For Dyninst, the leaked memory can be either (1) for useful states that are necessary for instrumentation and analysis, or (2) for intermediate states that are used for calculating (1). 

heaptrack lists all the significant call stacks that lead to the leaked memory, ranging from 1.71GB to 586MB (9.6GB in total). As I mentioned, not every leaked memory can be addressed:

1. 1.71GB leaked in `parseAPI/src/IA_IAPI.C:234`. This `IA_IAPI` class is solely for intermediate calculation, so all objects from this class should be freed. This is a leak to be fixed. 

2. 1.59GB leaked in `dyninstAPI/src/Parsing.C`. The related class `parse_block` is the Dyninst representation of the block for instrumentation. These objects are necessary for instrumentation, and so should be persistent. However, I found that this class contains member variables that are never used. So we have wasted memory when allocating `parse_block` objects.

3. 1.2GB leaked in `parseAPI/src/Function.C`. Here the memory is allocated when we add a basic block to a function's block map. This block map is necessary. 

4. 1.03GB leaked in `instructionAPI/srcInstructionDecoderImpl.C:128`. Here we construct RegisterAST. I don't know why the memory is leaked because the results are passed back as `boost::shared_ptr`. However, one flaw here is that we will always construct new RegisterAST for each instruction. This means that we will have enormous amount of Register AST for rax, rbx, .... . RegisterAST should not lead to 1.03GB memory usage. I think here we probably need a thread local map to avoid reconstruction of RegisterAST.

5. 1.01GB leaked in `parseAPI/src/Block.C`. Here the memory is allocated when adding source edges. There are some cleanup that I can do to shrink the memory usage here. Another thing to investigate here is how many sink edges have been allocated. An edge object costs about a dozens of bytes, and it takes additional memory in block's source edge and target edge list. We can just use a boolean in Block to represent whether this block has sink edge or not.

6. 946MB leaked in `parseAPI/src/Function.C:222`. The code is about finalizing functions. But it is not clear to me yet what and how memory is leaked.

7. 913MB leaked in  `instructionAPI/src/InstructionDecoderImpl.C:85`. Here we construct BinaryFunctionAST. This is similar to the RegisterAST leak.

8. 646MB leaked in `dataflowAPI/src/Absloc.C:285`. Here we construct Assignment, similar to the RegisterAST leak.

9. 568MB leaked in `parseAPI/src/Block.C` about edges.

**Comment by hainest** [comment]
@mxz297 There are some existing memory leak fixes in #317 that might be of use here. I was going to cherry-pick that PR into a new one, but I don't want to step on your work here.

**Comment by mxz297** [comment]
@hainest Those fixes in #317 are unlikely to be applicable. I updated the leaks with related changes that I have done. The peak memory usage went down from 14.5G to 10.8G. Will see if there are still any low hanging fruits.

**Comment by hainest** [comment]
@mxz297 Ok. I'll go ahead and cherry-pick 317.

**Comment by mxz297** [comment]
Closed via #758
